predicates-push-down.md

```diff
@@ -0,0 +1,151 @@
+| id                            | estRows  | task      | access object | operator info                                   |
++-------------------------------+----------+-----------+---------------+-------------------------------------------------+
+| Selection_7                   | 10000.00 | root      |               | isnull(test.s.a)                                |
+| └─HashJoin_8                  | 12500.00 | root      |               | left outer join, equal:[eq(test.t.a, test.s.a)] |
+|   ├─TableReader_13(Build)     | 10000.00 | root      |               | data:TableFullScan_12                           |
+|   │ └─TableFullScan_12        | 10000.00 | cop[tikv] | table:s       | keep order:false, stats:pseudo                  |
+|   └─TableReader_11(Probe)     | 10000.00 | root      |               | data:TableFullScan_10                           |
+|     └─TableFullScan_10        | 10000.00 | cop[tikv] | table:t       | keep order:false, stats:pseudo                  |
++-------------------------------+----------+-----------+---------------+-------------------------------------------------+
+6 rows in set (0.00 sec)
+```
+
+In this query，there is a predicate `s.a is null` on the inner table `s`。
+
+From the `explain` results，we can see that the predicate is not pushed below join operator. This is because the outer join will fill the inner table with `NULL` values when the on condition isn't satisfied, and the predicate `s.a is null` is used to filter the results after the join. If it is pushed down to the inner table below join, the execution plan is not equivalent with the original one.
+
+### Case 6: the predicates which contain user variables cannot be pushed down
+
+```sql
+create table t(id int primary key, a char);
+set @a = 1;
+explain select * from t where a < @a;
++-------------------------+----------+-----------+---------------+--------------------------------+
+| id                      | estRows  | task      | access object | operator info                  |
++-------------------------+----------+-----------+---------------+--------------------------------+
+| Selection_5             | 8000.00  | root      |               | lt(test.t.a, getvar("a"))      |
+| └─TableReader_7         | 10000.00 | root      |               | data:TableFullScan_6           |
+|   └─TableFullScan_6     | 10000.00 | cop[tikv] | table:t       | keep order:false, stats:pseudo |
++-------------------------+----------+-----------+---------------+--------------------------------+
+3 rows in set (0.00 sec)
+```
+
+In this query，there is a predicate `a < @a` on table `t`. The `@a` of predicate is a user variable.
+
+As can be seen from `explain` results, the predicate is not like case 2, which is simplified to `a < 1` and pushed down to TiKV. This is because the value of the user variable `@a` may changes during the computation, and TiKV is not aware of the changes. So TiDB will not replace `@a` with `1`, and will not be pushed down to TiKV.
```

Comment:
```suggestion
As can be seen from `explain` results, the predicate is not like case 2, which is simplified to `a < 1` and pushed down to TiKV. This is because the value of the user variable `@a` may change during the computation, and TiKV is not aware of the changes. So TiDB does not replace `@a` with `1`, and does not push down it to TiKV.
```